proxy/src/control/remote_stream.rs

```diff
@@ -0,0 +1,122 @@
+
+use futures::{Future, Poll, Stream};
+use http;
+use prost::Message;
+use std::fmt;
+use tower_h2::{Body, Data, RecvBody};
+use tower_grpc::{
+    self as grpc,
+    Streaming,
+    client::server_streaming::ResponseFuture,
+};
+
+/// Tracks the state of a remote response stream.
+///
+/// A remote may hold a `Receiver` that can be used to read `M`-typed messages from the
+/// remote stream.
+///
+/// If the Remote does not hold an active `Receiver`, `needs_reconnect()` returns true and
+/// `take_receiver()` returns `None`.
+#[derive(Debug)]
+pub struct Remote<M, F, B: Body = RecvBody>(Inner<M, F, B>);
+
+#[derive(Debug)]
+enum Inner<M, F, B: Body = RecvBody> {
```

Comment:
Why is it necessary to have separate `Remote` and `Inner` types? Usually we would have an `Inner` type if if we have an `Option<Inner>` that we sometimes `take()`, but that's not the case here. If we can combine this into `Remote`, let's do so, to make the code easier to read. Otherwise, let's add a comment here about why it is split out.